File: nocternal-playz/ios/NocternalKit/Sources/ThemeManager/GenreDetector.swift
import Foundation
import NocternalModel

public struct GenreMatch: Sendable {
    public let genre: GenreDefinition
    public let confidence: Double
    public let reason: String
}

/// Same detection rules as Kotlin `GenreDetector`: tag (aliases, then store tag families) → folder → specific
/// words in title/artist → podcast → tempo/energy. Text is normalised ("Hip-Hop/Rap" → "hip hop rap").
public struct GenreDetector: Sendable {
    private let aliases: [(String, GenreDefinition)]
    private let strong: [(String, GenreDefinition)]

    public static let genericWords: Set<String> = [
        "rain", "love", "chill", "dance", "party", "morning", "focus", "study", "sleep", "gym", "fitness", "running", "cardio",
        "motivation", "coding", "zen", "trap", "soul", "ballad", "romance", "romantic", "club", "feel good", "coffee", "sunrise",
        "acoustic", "folk", "guitar", "piano", "flute", "score", "episode", "news", "talk", "speech", "interview", "thunder",
        "house", "minimal", "acid", "industrial", "drone", "bedroom", "deep work", "concentration", "productivity", "evergreen",
        "disco", "slow jam", "healing", "yoga", "mindful", "lullaby", "ocean waves", "night sounds", "uplifting", "space music",
        "soundscape", "sitar", "retro bollywood", "golden era", "rap", "hip hop", "chillout", "chill out", "downtempo", "lounge",
    ]
    static let meaningless: Set<String> = ["other", "unknown", "misc", "miscellaneous", "genre", "none", "general", "various", "music", "<unknown>"]

    public init(genres: [GenreDefinition] = GenreCatalog.all) {
        var seen = Set<String>(), list: [(String, GenreDefinition)] = []
        for g in genres { for a in g.aliases { let n = GenreDetector.normalize(a); if seen.insert(n + "|" + g.id).inserted { list.append((n, g)) } } }
        aliases = list
        strong = list.filter { !GenreDetector.genericWords.contains($0.0) }
    }

    public static func normalize(_ s: String) -> String {
        s.lowercased().replacingOccurrences(of: #"[-_/\\.,;:+]"#, with: " ", options: .regularExpression)
            .replacingOccurrences(of: #"\s+"#, with: " ", options: .regularExpression).trimmingCharacters(in: .whitespaces)
    }

    public func detect(_ t: Track) -> GenreMatch? {
        if let raw = t.genreTag {
            let tag = GenreDetector.normalize(raw)
            if !tag.isEmpty && !GenreDetector.meaningless.contains(tag) {
                if let m = match(tag, aliases) { return GenreMatch(genre: m.1, confidence: 0.95, reason: "genre tag “\(raw)”") }
                if let g = family(tag, t) { return GenreMatch(genre: g, confidence: 0.8, reason: "genre tag “\(raw)”") }
            }
        }
        if t.isPodcast { return GenreMatch(genre: GenreCatalog.podcastMode, confidence: 0.9, reason: "podcast episode") }
        if let m = match(GenreDetector.normalize(t.folder), aliases) { return GenreMatch(genre: m.1, confidence: 0.8, reason: "folder “\(t.folder)”") }
        let text = GenreDetector.normalize([t.title, t.album, t.artist].joined(separator: " | "))
        if let m = match(text, strong) { return GenreMatch(genre: m.1, confidence: 0.7, reason: "“\(m.0)” in title/artist") }
        return byTempo(t)
    }

    private func family(_ tag: String, _ t: Track) -> GenreDefinition? {
        let e = t.energy ?? 0.5, bpm = t.bpm ?? 0
        func has(_ w: String...) -> Bool { w.contains { wordIndex(tag, $0) != nil } }
        if has("bollywood", "hindi", "filmi", "indian", "desi", "tamil", "telugu", "kannada", "malayalam", "marathi", "bengali", "bhojpuri", "gujarati", "tollywood", "kollywood") {
            if (t.year ?? 3000) < 2000 { return GenreCatalog.hindiClassics }
            return e > 0.65 || bpm >= 118 ? GenreCatalog.partyMix : GenreCatalog.romantic
        }
        if has("punjabi", "bhangra", "haryanvi") { return GenreCatalog.partyMix }
        if has("rock", "metal", "punk", "grunge", "alternative", "hardcore") { return GenreCatalog.workout }
        if has("electronic", "electronica", "electro", "dance", "club", "idm") { return e < 0.4 ? GenreCatalog.chillout : GenreCatalog.edm }
        if has("jazz", "blues", "bossa", "swing", "reggae", "easy listening") { return GenreCatalog.chillout }
        if has("classical", "ost", "film score", "orchestra", "opera", "baroque") { return GenreCatalog.instrumental }
        if has("country", "singer songwriter", "americana", "bluegrass") { return GenreCatalog.morningVibes }
        if has("k pop", "kpop", "j pop", "latin", "world", "afrobeat", "afrobeats") { return GenreCatalog.partyMix }
        if has("pop", "indie", "top 40") { return e > 0.6 || bpm >= 118 ? GenreCatalog.partyMix : GenreCatalog.morningVibes }
        if has("new age", "relaxation", "nature") { return GenreCatalog.meditation }
        if has("audiobook", "spoken", "comedy", "radio show") { return GenreCatalog.podcastMode }
        return nil
    }

    private func byTempo(_ t: Track) -> GenreMatch? {
        guard t.bpm != nil || t.energy != nil else { return nil }
        let e = t.energy ?? 0.5, b = t.bpm ?? 100
        let g: GenreDefinition
        switch true {
        case b >= 165 && e > 0.6: g = GenreCatalog.workout
        case (136...150).contains(b) && e > 0.6: g = GenreCatalog.trance
        case (122...136).contains(b) && e > 0.7: g = GenreCatalog.edm
        case (118...135).contains(b) && e > 0.55: g = GenreCatalog.partyMix
        case b < 70 && e < 0.3: g = GenreCatalog.sleep
        case b < 90 && e < 0.45: g = GenreCatalog.chillout
        case e >= 0.7: g = b > 140 ? GenreCatalog.workout : GenreCatalog.partyMix
        case e < 0.3: g = GenreCatalog.ambient
        case b < 100: g = GenreCatalog.lofi
        default: g = GenreCatalog.morningVibes
        }
        return GenreMatch(genre: g, confidence: 0.4, reason: "tempo \(Int(b)) BPM, energy \(Int(e * 100))%")
    }

    public func forMood(_ mood: Mood, hour: Int? = nil) -> GenreDefinition {
        switch mood {
        case .calm: return (hour ?? 0) >= 21 ? GenreCatalog.ambient : GenreCatalog.chillout
        case .energetic: return GenreCatalog.edm
        case .happy: return GenreCatalog.morningVibes
        case .melancholic: return GenreCatalog.lofi
        case .romantic: return GenreCatalog.romantic
        case .focused: return GenreCatalog.focus
        case .spiritual: return GenreCatalog.devotional
        case .sleepy: return GenreCatalog.sleep
        case .party: return GenreCatalog.partyMix
        }
    }

    private func match(_ text: String, _ index: [(String, GenreDefinition)]) -> (String, GenreDefinition)? {
        var best: (Int, String, GenreDefinition)?
        for (alias, g) in index {
            guard let i = wordIndex(text, alias) else { continue }
            if best == nil || i < best!.0 || (i == best!.0 && alias.count > best!.1.count) { best = (i, alias, g) }
        }
        return best.map { ($0.1, $0.2) }
    }

    private func wordIndex(_ text: String, _ alias: String) -> Int? {
        let chars = Array(text), a = Array(alias)
        guard !a.isEmpty, a.count <= chars.count else { return nil }
        var i = 0
        while i + a.count <= chars.count {
            if Array(chars[i..<(i + a.count)]) == a {
                let before: Character = i == 0 ? " " : chars[i - 1]
                let after: Character = i + a.count >= chars.count ? " " : chars[i + a.count]
                if !(before.isLetter || before.isNumber) && !after.isLetter { return i }
            }
            i += 1
        }
        return nil
    }
}

public enum TimeOfDayThemes {
    public static func genreForHour(_ hour: Int) -> GenreDefinition {
        switch ((hour % 24) + 24) % 24 {
        case 5...8: return GenreCatalog.morningVibes
        case 9...16: return GenreCatalog.focus
        case 17...20: return GenreCatalog.chillout
        case 21...23: return GenreCatalog.nightDrive
        default: return GenreCatalog.sleep
        }
    }
}

File: nocternal-playz/android/theme_manager/src/main/kotlin/com/nocternal/playz/theme/GenreDetector.kt
package com.nocternal.playz.theme

import com.nocternal.playz.model.GenreDefinition
import com.nocternal.playz.model.Mood
import com.nocternal.playz.model.Track

/** Result of genre detection with a confidence in 0..1 and a short reason for the UI/AI explanations. */
data class GenreMatch(val genre: GenreDefinition, val confidence: Float, val reason: String)

/**
 * Detects a built-in genre for a track, strongest evidence first:
 *  1. The genre tag, via the genre aliases ("Bhajan", "Psytrance").
 *  2. The genre tag, via common music-store tag families ("Bollywood", "Pop", "Rock", "Hip-Hop/Rap", "Jazz").
 *  3. The folder name (all aliases: "Music/Sleep" is intentional).
 *  4. Title, album and artist, using only specific aliases — ordinary words like "rain" or "love" in a title
 *     don't decide the genre ("Purple Rain" is not a sleep track).
 *  5. Podcast flag.
 *  6. Tempo and energy from on-device analysis, so every analysed song lands in a genre playlist.
 * Tags and aliases are normalised ("Hip-Hop/Rap" → "hip hop rap", "Lo-Fi" → "lo fi"). The alias that appears
 * first wins; on a tie the longer alias wins ("hard techno" over "techno").
 */
class GenreDetector(private val genres: List<GenreDefinition> = GenreCatalog.all) {

    private val aliasIndex: List<Pair<String, GenreDefinition>> =
        genres.flatMap { g -> g.aliases.map { normalize(it) to g } }.distinctBy { it.first to it.second.id }
    private val strongAliases = aliasIndex.filter { it.first !in GENERIC_WORDS }

    fun detect(track: Track): GenreMatch? {
        val tag = track.genreTag?.let(::normalize)?.takeIf { it.isNotBlank() && it !in MEANINGLESS_TAGS }
        if (tag != null) {
            match(tag, aliasIndex)?.let { return GenreMatch(it.second, 0.95f, "genre tag “${track.genreTag}”") }
            tagFamily(tag, track)?.let { return GenreMatch(it, 0.8f, "genre tag “${track.genreTag}”") }
        }
        if (track.isPodcast) return GenreMatch(GenreCatalog.PODCAST_MODE, 0.9f, "podcast episode")

        match(normalize(track.folder), aliasIndex)?.let { return GenreMatch(it.second, 0.8f, "folder “${track.folder}”") }
        val text = normalize(listOf(track.title, track.album, track.artist).joinToString(" | "))
        match(text, strongAliases)?.let { return GenreMatch(it.second, 0.7f, "“${it.first}” in title/artist") }

        return byTempo(track)
    }

    /** Fallback used when only a mood is known (AI mood detection). */
    fun forMood(mood: Mood, hour: Int? = null): GenreDefinition = when (mood) {
        Mood.CALM -> if (hour != null && hour >= 21) GenreCatalog.AMBIENT else GenreCatalog.CHILLOUT
        Mood.ENERGETIC -> GenreCatalog.EDM
        Mood.HAPPY -> GenreCatalog.MORNING_VIBES
        Mood.MELANCHOLIC -> GenreCatalog.LOFI
        Mood.ROMANTIC -> GenreCatalog.ROMANTIC
        Mood.FOCUSED -> GenreCatalog.FOCUS
        Mood.SPIRITUAL -> GenreCatalog.DEVOTIONAL
        Mood.SLEEPY -> GenreCatalog.SLEEP
        Mood.PARTY -> GenreCatalog.PARTY_MIX
    }

    /** Maps broad store/ID3 genres onto the built-in genres. */
    private fun tagFamily(tag: String, t: Track): GenreDefinition? {
        val energy = t.energy ?: 0.5f
        val bpm = t.bpm ?: 0f
        fun has(vararg w: String) = w.any { wordIndex(tag, it) >= 0 }
        return when {
            has("bollywood", "hindi", "filmi", "indian", "desi", "tamil", "telugu", "kannada", "malayalam", "marathi", "bengali",
                "bhojpuri", "gujarati", "tollywood", "kollywood", "hindustani film") -> when {
                (t.year ?: 3000) < 2000 -> GenreCatalog.HINDI_CLASSICS
                energy > 0.65f || bpm >= 118f -> GenreCatalog.PARTY_MIX
                else -> GenreCatalog.ROMANTIC
            }
            has("punjabi", "bhangra", "haryanvi") -> GenreCatalog.PARTY_MIX
            has("rock", "metal", "punk", "grunge", "alternative", "alt rock", "hardcore") -> GenreCatalog.WORKOUT
            has("electronic", "electronica", "electro", "dance", "club", "idm") -> if (energy < 0.4f) GenreCatalog.CHILLOUT else GenreCatalog.EDM
            has("jazz", "blues", "bossa", "swing", "reggae", "easy listening") -> GenreCatalog.CHILLOUT
            has("classical", "ost", "film score", "orchestra", "opera", "baroque") -> GenreCatalog.INSTRUMENTAL
            has("country", "singer songwriter", "americana", "bluegrass") -> GenreCatalog.MORNING_VIBES
            has("k pop", "kpop", "j pop", "latin", "world", "afrobeat", "afrobeats") -> GenreCatalog.PARTY_MIX
            has("pop", "indie", "top 40") -> if (energy > 0.6f || bpm >= 118f) GenreCatalog.PARTY_MIX else GenreCatalog.MORNING_VIBES
            has("new age", "relaxation", "nature") -> GenreCatalog.MEDITATION
            has("children", "kids", "nursery") -> GenreCatalog.MORNING_VIBES
            has("audiobook", "spoken", "comedy", "radio show") -> GenreCatalog.PODCAST_MODE
            else -> null
        }
    }

    /** Tempo/energy placement for analysed songs without useful tags. */
    private fun byTempo(t: Track): GenreMatch? {
        val bpm = t.bpm
        val energy = t.energy
        if (bpm == null && energy == null) return null
        val e = energy ?: 0.5f
        val b = bpm ?: 100f
        val g = when {
            b >= 165 && e > 0.6f -> GenreCatalog.WORKOUT
            b in 136f..150f && e > 0.6f -> GenreCatalog.TRANCE
            b in 122f..136f && e > 0.7f -> GenreCatalog.EDM
            b in 118f..135f && e > 0.55f -> GenreCatalog.PARTY_MIX
            b < 70f && e < 0.3f -> GenreCatalog.SLEEP
            b < 90f && e < 0.45f -> GenreCatalog.CHILLOUT
            e >= 0.7f -> if (b > 140f) GenreCatalog.WORKOUT else GenreCatalog.PARTY_MIX
            e < 0.3f -> GenreCatalog.AMBIENT
            b < 100f -> GenreCatalog.LOFI
            else -> GenreCatalog.MORNING_VIBES
        }
        return GenreMatch(g, 0.4f, "tempo ${b.toInt()} BPM, energy ${(e * 100).toInt()}%")
    }

    private fun match(text: String, index: List<Pair<String, GenreDefinition>>): Pair<String, GenreDefinition>? =
        index
            .mapNotNull { entry -> wordIndex(text, entry.first).takeIf { it >= 0 }?.let { it to entry } }
            .minWithOrNull(compareBy<Pair<Int, Pair<String, GenreDefinition>>> { it.first }.thenByDescending { it.second.first.length })
            ?.second

    /** Index of a whole-word match (so "rain" does not match inside "trance" or "chill" inside "chillhop"), or -1. */
    private fun wordIndex(text: String, alias: String): Int {
        if (alias.isEmpty()) return -1
        var from = 0
        while (true) {
            val i = text.indexOf(alias, from)
            if (i < 0) return -1
            val before = if (i == 0) ' ' else text[i - 1]
            val after = if (i + alias.length >= text.length) ' ' else text[i + alias.length]
            if (!before.isLetterOrDigit() && !after.isLetter()) return i
            from = i + 1
        }
    }

    companion object {
        fun normalize(s: String): String = s.lowercase().replace(Regex("[-_/\\\\.,;:+]"), " ").replace(Regex("\\s+"), " ").trim()

        /** Everyday words that only count as genre evidence in a genre tag or folder name, never in a title. */
        val GENERIC_WORDS = setOf(
            "rain", "love", "chill", "dance", "party", "morning", "focus", "study", "sleep", "gym", "fitness", "running", "cardio",
            "motivation", "coding", "zen", "trap", "soul", "ballad", "romance", "romantic", "club", "feel good", "coffee", "sunrise",
            "acoustic", "folk", "guitar", "piano", "flute", "score", "episode", "news", "talk", "speech", "interview", "thunder",
            "house", "minimal", "acid", "industrial", "drone", "bedroom", "deep work", "concentration", "productivity", "evergreen",
            "disco", "slow jam", "healing", "yoga", "mindful", "lullaby", "ocean waves", "night sounds", "uplifting", "space music",
            "soundscape", "sitar", "retro bollywood", "golden era", "rap", "hip hop", "chillout", "chill out", "downtempo", "lounge",
        )
        val MEANINGLESS_TAGS = setOf("other", "unknown", "misc", "miscellaneous", "genre", "none", "general", "various", "music", "<unknown>")
    }
}
